***Support Vector Machines***
    Using SVM with Python to predict whether a breast cancer tumor is malignant or benign. Exercise uses numpy, pandas, and scikitlearn and utilizes train test split, SVM, SVC,  and GridSearch to identify the best parameters for prediction.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# 1 point

# import and load the breast cancer dataset 
from sklearn import datasets
# output a variable where the data is loaded straight from sklearn



# YOUR CODE HERE
cancer = datasets.load_breast_cancer()


In [3]:
assert cancer.data.shape == (569,30)

In [4]:
print(cancer['DESCR'])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [5]:
cancer['feature_names']

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [13]:
# 1 point

#load the data from the dataset into a pd df

#output a dataframe with the data of the dataset and the column names as feature names
#IT SHOULD NOT INCLUDE THE TARGET 

df_feat = pd.DataFrame(cancer.data,columns=cancer.feature_names)
# YOUR CODE HERE


In [10]:
b = True
for i in cancer['feature_names']:
    if df_feat[i].shape != (569,):
        b = False
        break
assert b

In [17]:
cancer.target.shape

(569,)

In [21]:
# 1 point

# create a dataframe with all the targets and the column name your be 'Cancer'
# output a single column dataframe
df_target = pd.DataFrame(cancer.target, columns=['Cancer'])

# YOUR CODE HERE


In [22]:
assert df_target['Cancer'].shape == (569,)

In [23]:
#1 point
# get the train test split, tested with random state 0 and test_size 0.2
from sklearn.model_selection import train_test_split
X,y = cancer.data , cancer.target
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)
# all the subsets should be numpy arrays
# TIP carefull with the shape of the Y
# YOUR CODE HERE


In [27]:
assert x_train.shape == (455,30)
assert x_test.shape == (114,30)
assert y_train.shape == (455,)
assert y_test.shape == (114,)

In [30]:
# 2 points
#Train the support vector classifier
# import if needed
from sklearn.svm import SVC
clf = SVC()
clf.fit(x_train,y_train)

SVC()

In [31]:
assert clf.support_vectors_.shape == (123,30)

In [32]:
# 3 points
# predict the testing data and get the accuracy
from sklearn import metrics
# output preds is the array with the predictions and acc is a float between 0 and 1
preds = clf.predict(x_test)
acc = metrics.accuracy_score(y_test,preds)
print('Accuracy is:', acc)
# YOUR CODE HERE



Accuracy is: 0.9298245614035088


In [ ]:
from sklearn.metrics import confusion_matrix



conf_matrix = confusion_matrix(y_test, preds)
print(conf_matrix)

In [33]:
assert acc > 0.85
assert preds.shape == (114,)

***Everything is being classified into a single class! Use Grid Search to search for the best parameters! (Scikitlearn GridSearchCV)***


In [80]:
# 2 points
#Set parameters
# the parameters should be in a dict format
params = {'C':np.linspace(1,10,10).astype(int), 'kernel':['linear',  'sigmoid']      }
# YOUR CODE HERE


In [81]:
assert len(params) == 2
assert len(params['kernel']) == 2

In [82]:
#Fit grid (Note to self: first it runs the same loop with cross-validation to find the best parameter combination. Once it has the best combination, it runs fit again on all data passed to fit (without cross-validation), to built a single new model using the best parameter setting.)
# 4 points
# add an svm to the grid and train it
# the SVM should be an untrained one
from sklearn.model_selection import GridSearchCV
svc = SVC(random_state=0)
grid_svc = GridSearchCV(svc,params)
grid_svc.fit(x_train,y_train)

# YOUR CODE HERE



GridSearchCV(estimator=SVC(random_state=0),
             param_grid={'C': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10]),
                         'kernel': ['linear', 'sigmoid']})

In [83]:
tuned_model =grid_svc.best_estimator_
print(tuned_model)
print(grid_svc.best_score_)

SVC(C=1, kernel='linear', random_state=0)
0.9604395604395604


In [84]:
# pd.DataFrame(grid_svc.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0        1.321381      0.531537         0.001400    4.897264e-04       1   
1        0.009194      0.000980         0.002199    4.004241e-04       1   
2        1.056346      0.509440         0.000400    4.898042e-04       2   
3        0.009194      0.000979         0.001999    1.288343e-06       2   
4        2.595992      0.645078         0.001000    4.623108e-07       3   
5        0.015391      0.001959         0.003199    4.009010e-04       3   
6        3.732489      0.982939         0.001200    4.000427e-04       4   
7        0.008994      0.000895         0.002199    4.001644e-04       4   
8        4.648123      3.045752         0.001000    6.323345e-04       5   
9        0.009394      0.000489         0.002399    7.991793e-04       5   
10       4.591957      3.141175         0.000999    1.256174e-06       6   
11       0.009193      0.000979         0.003199    1.165370e-03       6   
12       5.574749      2.282611         0.001000    6.324857e-04       7   
13       0.011592      0.001624         0.003199    1.165289e-03       7   
14       4.400076      1.875949         0.000800    3.999473e-04       8   
15       0.008595      0.000800         0.002199    3.998998e-04       8   
16       6.191967      2.932839         0.001000    1.083166e-06       9   
17       0.009193      0.000980         0.002398    4.892591e-04       9   
18       5.614524      3.456338         0.001401    7.996798e-04      10   
19       0.011393      0.002058         0.002798    9.792188e-04      10   

   param_kernel                          params  split0_test_score  \
0        linear    {'C': 1, 'kernel': 'linear'}           0.989011   
1       sigmoid   {'C': 1, 'kernel': 'sigmoid'}           0.527473   
2        linear    {'C': 2, 'kernel': 'linear'}           0.967033   
3       sigmoid   {'C': 2, 'kernel': 'sigmoid'}           0.494505   
4        linear    {'C': 3, 'kernel': 'linear'}           0.967033   
5       sigmoid   {'C': 3, 'kernel': 'sigmoid'}           0.494505   
6        linear    {'C': 4, 'kernel': 'linear'}           0.978022   
7       sigmoid   {'C': 4, 'kernel': 'sigmoid'}           0.483516   
8        linear    {'C': 5, 'kernel': 'linear'}           0.967033   
9       sigmoid   {'C': 5, 'kernel': 'sigmoid'}           0.472527   
10       linear    {'C': 6, 'kernel': 'linear'}           0.967033   
11      sigmoid   {'C': 6, 'kernel': 'sigmoid'}           0.461538   
12       linear    {'C': 7, 'kernel': 'linear'}           0.967033   
13      sigmoid   {'C': 7, 'kernel': 'sigmoid'}           0.461538   
14       linear    {'C': 8, 'kernel': 'linear'}           0.967033   
15      sigmoid   {'C': 8, 'kernel': 'sigmoid'}           0.461538   
16       linear    {'C': 9, 'kernel': 'linear'}           0.945055   
17      sigmoid   {'C': 9, 'kernel': 'sigmoid'}           0.461538   
18       linear   {'C': 10, 'kernel': 'linear'}           0.956044   
19      sigmoid  {'C': 10, 'kernel': 'sigmoid'}           0.461538   

    split1_test_score  split2_test_score  split3_test_score  \
0            0.967033           0.912088           0.967033   
1            0.428571           0.505495           0.483516   
2            0.956044           0.901099           0.967033   
3            0.417582           0.483516           0.450549   
4            0.967033           0.901099           0.956044   
5            0.395604           0.483516           0.461538   
6            0.967033           0.912088           0.956044   
7            0.395604           0.483516           0.450549   
8            0.967033           0.901099           0.967033   
9            0.395604           0.483516           0.439560   
10           0.967033           0.901099           0.956044   
11           0.406593           0.483516           0.439560   
12           0.967033           0.901099           0.956044   
13           0.406593           0.483516           0.4

In [85]:
dic = grid_svc.get_params(deep=False)
assert type(dic['estimator']) == type(SVC()) 
assert dic['pre_dispatch'] == '2*n_jobs'

In [86]:
# 2 point
#Get in a list the best parameters of the grid
#output python list with the best parameters sorted
best_params = sorted(grid_svc.cv_results_)



In [87]:
assert best_params[0] == 'mean_fit_time'
assert best_params[1] == 'mean_score_time'
assert best_params[2] == 'mean_test_score'
assert best_params[3] == 'param_C'

In [88]:
# 5 points
# predict the testing set and get the accuracy
# accuracy as always will be between 0 and 1

best_model = tuned_model
best_model.fit(x_train,y_train)
preds = best_model.predict(x_test)
acc = metrics.accuracy_score(y_test,preds)
print('Accuracy is:', acc)

# YOUR CODE HERE


Accuracy is: 0.956140350877193


In [89]:

assert acc > 0.92
assert preds[preds.argmax()] == 1
assert preds.shape == (114,)

In [90]:
# 3 points
# get the confusion matrix
#import if needed




conf_mat = confusion_matrix(y_test, preds)
print(conf_mat)
# YOUR CODE HERE




[[46  1]
 [ 4 63]]


In [92]:
assert conf_mat.shape == (2,2)
assert conf_mat[0][1] < 7
assert conf_mat[1][0] < 3


AssertionError: 

# Discussion of results

### Using the default values of SVC to train the data an accuracy of about 92% was gotten 
### But performing a gridsearch to get the best parameters to use for the clf, I got the best paramter to be (C=1, kernel='linear', random_state=0)
### Using the above to train the model, there was an increase in the accuracy from 92% to about 96%